docs/refman/ppr2samba.8.pod
#
# mouse:~ppr/src/docs/refman/ppr2samba.8.pod
# Copyright 1995--2006, Trinity College Computing Center.
# Written by David Chappell.
#
# This file is part of PPR.  You can redistribute it and modify it under the
# terms of the revised BSD licence (without the advertising clause) as
# described in the accompanying file LICENSE.txt.
#
# Last modified 22 August 2006.
#

=head1 NAME

ppr2samba - Generates Samba configuration files from PPR configuration files

ppd2windrv - Converts PPD files and modifies Samba 2.2.x Configuration

=head1 SYNOPSIS

B<ppr2samba> [I<--nocreate>] [I<--debug>] [I<--version>] [I<--help>]

B<ppd2windrv>

=head1 DESCRIPTION

[This manpage is somewhat out of date.  An attempt has been made to correct
some of the more glaring problems, but it needs a thorough going over.]

These program are designed to make it easier to use PPR with Samba.  Samba
is an open source file and print server for SMB clients such as Microsoft
Windows 95, 98,  NT, etc.  For information on Samba refer to:

	http://www.samba.org/

The program B<ppr2samba> is used to export PPR's list of printers and groups
to a file which Samba can use the share them.  This file is called
B</etc/ppr/smb-include.conf>.

The program B<ppd2win95drv> can be used to convert PPR's collection of PPD
files to the correct format for Microsoft Windows.  Together with Samba 2.2.x
it allows not only Windows 95 clients, but Windows NT (including 2000 and XP)
clients to automatically download printer drivers.  The converted files are
stored in various subdirectories of F</var/spool/ppr/drivers>.  Since
B<ppd2win95drv> is a Perl script, you must have Perl 5 installed to use it.

If you add or remove printers from PPR, change their PPD files, or change
their descriptive comment strings, it is necessary to rerun B<ppr2samba>
so that Samba will perceive the change.

If you add, remove, replace, or modify any of the files listed in the B<[PPDs]>
section of F</etc/ppr/ppr.conf> or add or remove any of the Microsoft Windows
PostScript driver files in F</var/spool/ppr/drivers/>, or set a printer to use
a PPD file which is not in one of the directories listed in the B<[PPDs]>
section of F<ppr.conf>, you must rerun B<ppd2windrv> to reconfigure the
MS-Windows drivers.

=head2 Versions of Samba

It is recomended that you use that you use Samba 1.9.18 or later.  If you use a
version of Samba earlier than that, you must move the "printing = bsd" line in
the B<[pprproto]> section into the B<[global]> section.  You will also find
that some of the other configuration lines have no effect.

For fully automatic driver installation, Samba version 2.2.x is required.

=head2 Changes to Samba's Configuration

To use PPR with Samba you must add two lines to Samba's configuration file.
Samba's configuration file is called F<smb.conf>.  If you installed Samba from
source, it is likely to be found in the directory F</usr/local/samba/lib>.
Otherwise, it is probably F</etc/smb.conf> or F</etc/samba/smb.conf>.  Here are
the lines:

 include = /usr/lib/ppr/smb-protos.conf
 include = /etc/ppr/smb-include.conf

These lines can go at the end of the file or anywhere between sections but not
before the B<[global]> section.  The first line includes a file defines some
prototype shares which the actual print queue shares use as a source for
default values.  These prototype shares are B<[pprproto]>, and
B<[pprproto_pprpopup]>.  It also defines two disk shares which PPR uses.

The second line includes the file F<smb-include.conf> which is generated by
B<ppr2samba>.

Once you have added these lines you should run B<ppr2samba> to create the
initial F<smb-include.conf>.

You should add the name of Samba's guest user to the file
F</etc/ppr/acl/pprprox.allow>.  (The name of the user Samba uses for
guest logins is set in F<smb.conf>.  The default is generally "nobody".)
If you don't do this, jobs submitted
by guest users will appear in the MS-Windows print manager under the
name of the Samba guest user and Print Manager will grey out the
B<Document/Delete Document> menu item.

In versions of Samba at least as late as 2.0.x you must also
copy the "printer driver file =" line from the B<[pprprotos]> section
of F<smb-protos.conf> to the B<[global]> section of F<smb.conf>.  Hopefully
we can get a future version of Samba patched so that it is adaquate
to have this line in F<smb-protos.conf> with the others.

=head2 Contents of smb-protos.conf

When PPR is installed, it installs the lastest version of F<smb-protos.conf>
as F<smb-protos.conf.sample>.  If you don't have a F<smb-protos.conf> then
the sample file is copied to the real one when you run fixup.  When you upgrade
PPR you should compare the new sample file to the one you are using and
incorporate any needed changes.

The B<[pprproto]> section defines commands to submit jobs, delete jobs, list
the queue, and so forth.  When B<ppr2samba> generates the F<smb-include.conf>
file it will place a "copy =" line in each section it generates.  By default
these lines will read "copy = pprproto", so the B<[pprproto]> section serves
to set default values for PPR shares.

The option "B<-X %m@samba>" in the "B<print command =>" and "B<lprm command =>"
lines make sure that a user can only delete his own jobs.

The option "B<-f %m>" in the print command causes the name of the sending machine to appear
as the user on banner pages and in queue listings.  This option will be ignored
unless the user name which Samba uses to run B<ppr> under is listed in
F</etc/ppr/acl/pprprox.allow>.  That is why it is recomended that you put
the name of the Samba guest user in F<pprprox.allow>.  That way, guest users
can identify themselves for job tracking purposes, but other users will not
be able to override their identification.

The options "B<-e responder -m samba -r %m>" cause error messages to be sent back to the
client through the use of the responder B<samba>.  The responder B<samba> uses
Samba's B<smbclient> to send a message to the clients messenger service (winpopup
on Windows 95).

The "B<-U>" option causes the temporary file created by Samba to be deleted
after B<ppr> has copied it into the queue.

The option "B<-C ''>" prevents the job title from defaulting to the file name
since the name of the temporary file is not generally meaningful.

The print command is put into the background because it can take a long time
to run if certain filters are invoked.

The section B<[pprproto_pprpopup]> is an alternative prototype for PPR shares.
If jobs are submitted to a share which uses this prototype, then the wrapper
script B</usr/lib/ppr/samba_submitter> will be used instead of invoking
B<ppr> directly.  Before using B<ppr> to submit the job, this wrapper script
will attempt to contact a program called B<pprpopup.tcl> which is hopefully
running on the client.  The program B<pprpopup.tcl> is a Tcl/Tk script for
use on computers running Windows 95 or NT in public computer rooms.  It can
ask users to enter their names and it can receive messages which inform users
about what happened to their jobs.

If B<pprpopup.tcl> is running on the client, then it will pop up a
window which asks the user for his name.  The wrapper script will receive
the response and use it as the argument for the B<-f> switch when it invokes
B<ppr>.  It will also use the B<-m> switch to select the B<pprpopup> responder.
If B<pprpopup.tcl> is not found to be running on the client, then printing
proceeds normally and the B<samba> responder is selected.

To set a printer or group up to use B<pprpopup.tcl>, you must select the
prototype B<[pprproto_pprpopup]> instead of B<[pprproto]>.  To do this, add
this line to the printer or group configuration file:

 ppr2samba-prototype: pprproto_pprpopup

You must rerun B<ppr2samba> for the change to take effect.

The share B<[pprwin95]> contains driver files for downloading by
Microsoft Windows 95 computers.

The share B<[pprclipr]> is used by the B<audio> responder and B<audio>
commentator when sending a sound file to B<pprpopup.tcl> on a computer
running Microsoft Windows 95.

=head2 What ppr2samba Does

This program B<ppr2samba> reads the PPR printer and group configuration
files and generates the file F</etc/ppr/smb-include.conf>.  The file will
contain one Samba printer share definition for each PPR printer or group.
(See below for how to exclude printers or groups.)

Since it can only perform one function, and that non-disruptive one, any user
may run B<ppr2samba>.

Versions of B<ppr2samba> before 1.40 also generated a file called
F</etc/ppr/printcap>.  This was for compatibility with an older and
inferior scheme for using Samba and PPR together.

=head2 Special Configuration Lines

When B<ppr2samba> is reading printer and group configuration files it
recognizes several special configuration lines which only it pays
attention to.

One such line is "C<ppr2samba:>".  This takes a numberic argument.  If the
argument is non-zero, the printer or group is included in the files
generated by B<ppr2samba>, if it is zero it is ommited.  If no such line is
present, then the printer or group is included.  For example, to exclude
a printer from sharing by Samba, we add this line to its configuration file:

 ppr2samba: 0

Another such line is "C<ppr2samba-prototype:>".  This line takes a string
argument.  This argument is the name of another Samba share that should be
used as a prototype for this printer share.  By default this is "pprproto",
a prototype share defined in F<smb-protos.conf>.

To make a printer share use the B<[pprproto_pprpopup]> prototype so that
the B<pprpopup.tcl> program running on the client will be contacted to ask
the user for his name, put this line in the printer or group configuration file:

 ppr2samba-prototype: pprproto_pprpopup

The "C<ppr2samba-prototype:>" line was added in PPR version 1.41.  In
previous versions, the only way to specify the prototype was as a second
argument to the "C<ppr2samba:>" line, like this:

 ppr2samba: 1 pprproto_pprpopup

This obsolete format is still supported for compatibility.

Another such line is "C<ppr2samba-drivername:>" which can be used to override
the value B<ppr2samba> would select for the "printer driver =" line in
F<smb-include.conf>.  For example, to cause clients to use the PPD file
for the imaginary printer called "PPR Generic", put this line in a printer
or group configuration file:

 ppr2samba-drivername: PPR Generic

(Note that the printer PPR Generic is imaginary only in the sense that no
hardware with that name actually exists.  However, a PPD file describing this
hypothetical printer does exist.  This PPD file can be used to configure an
MS-Windows or Macintosh print driver so that it can invoke certain special
features of the PPR feature such as n-up printing.)

Prior to PPR version 1.40, B<ppr2samba> looked for a "C<ms-driver-name:>" line.
The name of this line was changed to "C<ppr2samba-drivername:>" in PPR
version 1.40.  If any of the old lines are found, B<ppr2samba> will print
warnings but will otherwise ignore them.

People who want to share many print queues might also want to know about the
"ppr2samba-vserver:" line.  This takes a numberic argument in the range
1 through 10.  If this line is present in a printer's or a group's
configuration file, then in addition to writing a section in F<smb-include.conf>,
B<ppr2samba> will write a section in one of ten files with names in the
form F<smb-include-N.conf>.  This will allow you to split your Samba print
server up into a number of `virtual servers'.  Some people want to do this
because under some circumstances there are limits on the size of a single
server's browse list.  [Unfortunately, the details on how to do this are
currently missing from this document.  If you are one of the people who
requested this feature, please write to <ppr-bugs@trincoll.edu> with
a description of how to set up F<smb.conf> for this mode of operation.]

=head2 Setting Configuration Lines

If you do not wish to edit the queue's configuration file directly, you can use
B<ppad>.  For example, to set the "ppr2samba-drivername:" line for the printer
"myprn":

	$ ppad addon ppr2samba-drivername "PPR Generic"

If you later want to remove the option, set it without supplying a value:

	$ ppad addon ppr2samba-drivername

See the B<ppad> man page for general descriptions of the B<ppad addon>, B<ppad group addon>
and B<ppad alias addon> commands.

=head2 Command Line Options for B<ppr2samba>

The B<ppr2samba> command has an option called B<--nocreate>.  When this option
is used it will do nothing if B<smb-include.conf> does already exist.  This
option is used by the PPR WWW administration tools to skip the update of
B<smb-include.conf> if the Samba tie-in hasn't been set up.

The B<--version> option prints the PPR version number.

The B<--debug> option turns on messages which describe what B<ppr2samba> is doing.

The B<--help> option prints a brief description of the command line options
that B<ppr2samba> accepts.

=head2 What ppd2win95drv does.

The program B<ppd2win95drv> is a Perl script which prepares the directory
F</var/spool/ppr/drivers/WINPPD/> and F</var/spool/ppr/drivers/WIN40/> as a
directories from which to dispense printer drivers for Microsoft Windows 95.

Setting up this feature is not simple.  The reason is that it requires files
which may not be distributed with PPR.  These files are components of the
PostScript driver that comes with Microsoft Windows 95, and the Adobe
PostScript driver versions 4.1 and 4.2x.  You must obtain the files for
the Windows 95 driver or the Adobe 4.1 driver or both.  You may also
choose to obtain the files for Adobe's 4.2x driver which works with
PostScript level 2 printers only.

The driver that comes with Windows 95 is the easiest to get.  Install the
driver for a PostScript printer on a Windows 95 computer and copy the
following files from the F<SYSTEM> directory to F</var/spool/ppr/drivers/WIN40>
on the PPR server:

	PSCRIPT.DRV
	PSCRIPT.HLP
	FONTS.MFM
	ICONLIB.DLL
	PSMON.DLL
	PSCRIPT.INI

When this is done, run B<ppd2win95drv>.  Drag-and-drop printer installation
should work now.

Unfortunately, the PostScript driver that comes with Windows 95 has some
annoying bugs.  Adobe has distributed several later versions that fixes many
of them.  Version 4.1 from Adobe works with PostScript level 1 and level 2
printers.  To obtain it, go to:

http://www.adobe.com/supportservice/custsupport/LIBRARY/pdrvwin.htm

where you will find a link to the driver.  While you are there you might as
well download the 4.2x driver too.  (Note, it appears that Adobe is no
longer distributing the version 4.1 driver.  This is a pity.  Does anyone
know of a place from which it may still be downloaded?)

Install the driver on a computer running Windows 95 and then copy these files
from the F<SYSTEM> directory to F</var/spool/ppr/drivers/WIN40> on the PPR server:

	ADOBEPS4.DRV (rename to ADBEPS41.DRV)
	ADOBEPS4.HLP
	FONTS.MFM
	ICONLIB.DLL
	PSMON.DLL

No you can rerun B<ppd2win95drv> and now when you drag-and-drop a PPR printer
share, the newer Adobe driver will be installed instead of the one that
came with Windows 95.

The 4.2x Adobe driver is newer than the one we installed above, but it won't
work with PostScript level 1 printers.  Adobe says that it is not possible
to have the 4.1 and the 4.2x drivers installed simulaineously on the same
Windows 95 computer because the drivers both have a file called F<ADOBEPS4.DRV>.

History.txt
mouse:~ppr/src/History.txt

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of
reading and setting terminal characteristics instead of the ioctl() method.


samba/ppr2samba.c
/*
** mouse:~ppr/src/samba/ppr2samba.c
** Copyright 1995--2004, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
**
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 17 December 2004.
*/

/*
** This program scans PPR's printer and group configuration files
** and generates a file for Samba to include from smb.conf.
*/

#include "before_system.h"
#include <sys/stat.h>
#include <errno.h>
#include <dirent.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <stdlib.h>
#include <time.h>
#include <ctype.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "util_exits.h"
#include "queueinfo.h"
#include "version.h"

/* Here we define the names of the files we use. */
const char smb_include_conf[] = CONFDIR"/smb-include.conf";
const char smb_include_x_conf[] = CONFDIR"/smb-include-%d.conf";
const char smb_protos_conf[] = HOMEDIR"/lib/smb-protos.conf";

/* And here are the directories we search. */
struct DIRS {
	const char *name;
	enum QUEUEINFO_TYPE type;
	} dirs[] = {
	{ALIASCONF, QUEUEINFO_ALIAS},
	{GRCONF, QUEUEINFO_GROUP},
	{PRCONF, QUEUEINFO_PRINTER},
	{NULL, QUEUEINFO_SEARCH}
	};

/* The files we will be sending our output to: */
#define MAX_VSERVERS 10
FILE *outfiles[MAX_VSERVERS + 1];

/* Should be be verbose? */
int debug = FALSE;

void warning(const char *message, ...)
#ifdef __GNUC__
	__attribute__ (( format (printf, 1, 2) ))
#endif
	;
void warning(const char *message, ... )
	{
	va_list va;
	va_start(va, message);
	fputc('\n', stderr);
	fputs(_("Warning: "), stderr);
	vfprintf(stderr, message, va);
	fputc('\n', stderr);
	fputc('\n', stderr);
	va_end(va);
	} /* end of warning() */

/*
** Create one of the files that will be included from smb.conf.
*/
static FILE *create_output_file(int vserver)
	{
	char fname[MAX_PPR_PATH];
	const char *fnp;
	FILE *f;
    static const char *timestamp_string = NULL;

	if(vserver == 0)
		{
		fnp = smb_include_conf;
		}
	else
		{
		ppr_fnamef(fname, smb_include_x_conf, vserver);
		fnp = fname;
		}

	if((f = fopen(fnp, "w")) == (FILE*)NULL)
		gu_Throw(_("Can't open \"%s\", errno=%d (%s)\n"), fnp, errno, gu_strerror(errno));

	if(!timestamp_string)
		{
		time_t time_now;
		char *temp;
		time(&time_now);
		temp = ctime(&time_now);
		timestamp_string = gu_strndup(temp, strcspn(temp, "\n"));
		}

	/* Put an explanatory header on the file: */
	fprintf(f,
	";\n"
	"; %s\n"
	"; %s\n"
	";\n"
	"; This file was generated by ppr2samba.  Do not edit it.\n"
	"; To update it, rerun ppr2samba.  For furthur information,\n"
	"; including how to exclude printers and groups from this file,\n"
	"; consult the ppr2samba(8) man page.\n"
	";\n\n", fnp, timestamp_string);

	return f;
	}

/*
** This is called from emmit_record(), once for the main file, and once
** for the virtual server file, if a virtual server entry was requested.
*/
static void write_smb_conf_record(const char directory[], const char name[], const char comment[], const char drivername[], const char proto[], int vserver)
	{
	FILE *f;

	if(!outfiles[vserver])
		outfiles[vserver] = create_output_file(vserver);

	f = outfiles[vserver];

	/* Emmit the smb.conf section. */
	fprintf(f, "; %s/%s\n", directory, name);
	fprintf(f, "[%s]\n", name);
	fprintf(f, "  comment = %s\n", comment ? comment : "");
	fprintf(f, "  printer = %s\n", name);
	if(drivername)
		fprintf(f, "  printer driver = %s\n", drivername);
	fprintf(f, "  copy = %s\n", proto ? proto : "pprproto");
	fprintf(f, "  browseable = yes\n\n");
	} /* end of write_smb_conf_record() */

static void emmit_record(const char directory[], const char name[], const char comment[], const char drivername[], const char proto[], int vserver)
	{
	write_smb_conf_record(directory, name, comment, drivername, proto, 0);
	if(vserver != 0)
		write_smb_conf_record(directory, name, comment, drivername, proto, vserver);
	} /* end of emmit_record() */

/*
** Search the printer configuration file directory and make an
** entry in the printcap and smb.conf file for each printer found there.
**
** Rememer the drivername of each printer because we will need
** it to determine the drivername of any group which has it as
** a member.
*/
static void do_config_file(struct DIRS *dir, const char qname[], int *total_exported, int *total_error)
	{
	char fname[MAX_PPR_PATH];
	FILE *f = NULL;
	char *line = NULL;
	int line_len = 128;
	int exported = 1;					/* boolean */
	char *p;
	char *drivername = NULL;
	char *proto = NULL;
	int vserver = 0;
	void *qobj = NULL;

	if(debug)
		printf("    %s", qname);
	
	gu_Try {
		ppr_fnamef(fname, "%s/%s", dir->name, qname);
		if(!(f = fopen(fname, "r")))
			gu_Throw(_("can't open \"%s\", errno=%d (%s)"), fname, errno, gu_strerror(errno));
	
		while((line = gu_getline(line, &line_len, f)))
			{
			if((p = lmatchp(line, "ppr2samba:")))
				{
				char *temp = NULL;
				if(gu_sscanf(p, "%d %S", &exported, &temp) < 1)
					warning(_("queue has an invalid \"%s\" line"), "ppr2samba");
				else
					{
					if(proto)
						gu_free(proto);
					proto = temp;
					}
				}
			else if((p = lmatchp(line, "ppr2samba-prototype:")))
				{
				char *temp = NULL;
				if(gu_sscanf(p, "%S", &temp) < 1)
					warning(_("queue has an invalid \"%s\" line"), "ppr2samba-prototype");
				else
					{
					if(proto)
						gu_free(proto);
					proto = temp;
					}
				}
			else if((p = lmatchp(line, "ppr2samba-drivername:")))
				{
				char *temp;
				if(gu_sscanf(p, "%A", &temp) != 1)
					warning(_("queue has an invalid \"%s\" line"), "ppr2samba-drivername");
	
				else
					{
					if(drivername)
						gu_free(drivername);
					drivername = p;
					if(debug)
						printf("  drivername = %s\n", drivername);
					}
				}
			else if((p = lmatchp(line, "ppr2samba-vserver:")))
				{
				warning("the option \"ppr2samba-vserver\" is obsolete and will be removed");
				vserver = atoi(p);
				if(vserver < 1 || vserver > MAX_VSERVERS)
					{
					warning(_("queue has an invalid \"%s\" line"), "ppr2samba-vserver");
					vserver = 0;
					}
				}
			}
	
		if(exported)
			{
			qobj = queueinfo_new_load_config(dir->type, qname);

			emmit_record(dir->name, qname,
				queueinfo_comment(qobj),
				drivername ? drivername : queueinfo_shortNickName(qobj),
				proto,
				vserver
				);

			if(debug)
				printf(" exported");
			(*total_exported)++;
			}
		else if(debug)
			{
			printf(" not exported");
			}
		}
	gu_Final {
		if(debug)
			printf("\n");
		if(line)
			gu_free(line);
		if(f)
			fclose(f);
		if(proto)
			gu_free(proto);
		if(drivername)
			gu_free(drivername);
		if(qobj)
			queueinfo_free(qobj);
		}
	gu_Catch {
		switch(dir->type)
			{
			case QUEUEINFO_ALIAS:
				gu_wrap_eprintf(_("Skipping alias \"%s\": %s\n"), qname, gu_exception);
				break;
			case QUEUEINFO_GROUP:
				gu_wrap_eprintf(_("Skipping group \"%s\": %s\n"), qname, gu_exception);
				break;
			case QUEUEINFO_PRINTER:
				gu_wrap_eprintf(_("Skipping printer \"%s\": %s\n"), qname, gu_exception);
				break;
			case QUEUEINFO_SEARCH:
				gu_Throw("assertion failed");
			}
		(*total_error)++;
		}
	} /* end of do_printers() */

/*
** Look over the Samba configuration and point out problems.
*/
static void check_for_problems(void)
	{
	FILE *f;
	char *line = NULL;
	int line_space = 80;
	gu_boolean saw_pprproto = FALSE;
	gu_boolean saw_pprprnt = FALSE;
	char *si, *di;

	if(!(f = fopen(smb_protos_conf, "r")))
		{
		fprintf(stderr, _("Can't open \"%s\", errno=%d (%s)\n"), smb_protos_conf, errno, gu_strerror(errno));
		return;
		}

	while((line = gu_getline(line, &line_space, f)))
		{
		for(si=di=line; *si; si++)
			{
			if(!isspace(*si))
				*(di++) = *si;
			}
		*di = '\0';

		if(strncmp(line, "[pprproto]", 10) == 0)
			saw_pprproto = TRUE;
		else if(strncasecmp(line, "[pprprnt$]", 10) == 0)
			saw_pprprnt = TRUE;
		}

	fclose(f);

	if(!saw_pprproto)
		{
		warning(_("The \"%s\" file doesn't contain a [pprproto] section."), smb_protos_conf);
		}

	if(!saw_pprprnt)
		{
		warning(_("The \"%s\" file doesn't contain a [pprprnt$] section."), smb_protos_conf);
		}

	} /* end of check_for_problems() */

static void help(FILE *outfile)
	{
	fprintf(outfile, _("Usage: %s\n"), "ppr2samba [--debug] [--nocreate] [--version] [--help]");
	fprintf(outfile, _("\t--debug      explain what is being done\n"));
	fprintf(outfile, _("\t--nocreate   don't create %s if it doesn't exist\n"), smb_include_conf);
	fprintf(outfile, _("\t--version    print the PPR version number\n"));
	fprintf(outfile, _("\t--help       print this message\n"));
	}

int main(int argc, char *argv[])
	{
	int x;
	gu_boolean opt_create = TRUE;
	int total = 0, total_exported = 0, total_errors = 0;

	/* Initialize internation messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	umask(PPR_UMASK);

	for(x=1; x < argc; x++)
		{
		if(strcmp(argv[x], "-d") == 0 || strcmp(argv[x], "--debug") == 0)
			{
			debug = TRUE;
			continue;
			}
		if(strcmp(argv[x], "--version") == 0)
			{
			puts(COPYRIGHT);
			puts(VERSION);
			return EXIT_OK;
			}
		if(strcmp(argv[x], "--nocreate") == 0)
			{
			opt_create = FALSE;
			continue;
			}
		if(strcmp(argv[x], "--help") == 0)
			{
			help(stdout);
			return EXIT_OK;
			}

		help(stderr);
		return EXIT_SYNTAX;
		}

	if(debug)
		printf("ppr2samba\n" VERSION "\n\n");

	gu_Try {
		struct DIRS *dir;
		DIR *dirobj;
		struct dirent *direntp;
		int len;
	
		/* If we shouldn't create smb-include.conf, make sure it exists. */
		if(! opt_create)
			{
			struct stat statbuf;
			if(stat(smb_include_conf, &statbuf) != 0)
				{
				/* If the error is something other than file not found,
				   then say we couldn't open it since that is what would
				   have been true if we had tried. */
				if(errno != ENOENT)
					gu_Throw(_("Can't open \"%s\", errno=%d (%s)"), smb_include_conf, errno, gu_strerror(errno));
	
				/* File not found, say so and get out since we
				   have been asked not to create it. */
				gu_wrap_printf(_("Absence of \"%s\" indicates that export to Samba is not desired.\n"), smb_include_conf);
				return EXIT_OK;
				}
			}
	
		/* Clear the file array. */
		for(x=0; x <= MAX_VSERVERS; x++)
			outfiles[x] = NULL;

		/* Do aliases, groups, and printers. */
		for(dir=dirs; dir->name; dir++)
			{
			if(debug)
				gu_wrap_printf("Searching directory \"%s\"...\n", dir->name);
	
			if(!(dirobj = opendir(dir->name)))
				gu_Throw(_("Can't open directory \"%s\", errno=%d (%s)\n"), dir->name, errno, gu_strerror(errno));
	
			while((direntp = readdir(dirobj)))
				{
				/* Skip . and .. and hidden files. */
				if(direntp->d_name[0] == '.')
					continue;
	
				/* Skip Emacs style backup files. */
				len = strlen(direntp->d_name);
				if(len > 0 && direntp->d_name[len-1] == '~')
					continue;

				total++;
				do_config_file(dir, direntp->d_name, &total_exported, &total_errors);
				}
	
			closedir(dirobj);
			}
	
		/* Close the smb.conf fragment files and delete those
		   we didn't use. */
		for(x=0; x < MAX_VSERVERS; x++)
			{
			if(outfiles[x])
				{
				fclose(outfiles[x]);
				}
			else
				{
				char fname[MAX_PPR_PATH];
				ppr_fnamef(fname, smb_include_x_conf, x);
				unlink(fname);
				}
			}
	
		/* Report on what we did. */
		if(total == 1)
			gu_wrap_printf(_("%d of 1 queue exported to \"%s\".\n"), total_exported, smb_include_conf);
		else
			gu_wrap_printf(_("%d of %d queues exported to \"%s\".\n"), total_exported, total, smb_include_conf);
		if(total_errors > 0)
			gu_wrap_printf(_("%d of %d queue(s) skipt due to errors.\n"), total_errors, total);
	
		/* Produce warning messages. */
		check_for_problems();
	
		if(debug)
			printf(_("Done.\n"));
		}
	gu_Catch {
		fprintf(stderr, "%s: %s\n", argc > 0 ? argv[0] : "?", gu_exception);
		return EXIT_INTERNAL;
		}
	
	return EXIT_OK;
	} /* end of main() */

/* end of file */

